### ISSUE TITLE
nix-channel --update: "No space left on device" despite having space

### ISSUE THREAD

**Issue by golddranks** [NixOS/nix#8732]
**Describe the bug**

I'm on NixOS 22.11. I found that `nix-channel --update` started failing with the following behaviour:

```
# nix-channel --update
unpacking channels...
error: creating directory '/tmp/nix-build-nixos-22.11.drv-1': No space left on device
error: program '/nix/store/4mkr15glvxawmvmadj7vhxpqndwb6rpg-nix-2.11.1/bin/nix-env' failed with exit code 1
```

This seems strange, as `df` is showing that there should be enough space, especially given that it's just a channel update:
```
# df -h
Filesystem      Size  Used Avail Use% Mounted on
devtmpfs        396M     0  396M   0% /dev
tmpfs           3,9G     0  3,9G   0% /dev/shm
tmpfs           2,0G  7,0M  2,0G   1% /run
tmpfs           3,9G  408K  3,9G   1% /run/wrappers
/dev/sdd1       102G   65G   32G  67% /
/dev/sdc1       3,6T  2,9T  520G  85% /mnt/Valtavuus
/dev/sdd3       510M  247M  264M  49% /boot
/dev/sda2       2,8T  2,2T  600G  79% /mnt/Avaruus
/dev/sda2       2,8T  2,2T  600G  79% /srv/samba/Filesaari/Musiikki
tmpfs           792M     0  792M   0% /run/user/1000
```

My nix partition is on `/`:

```
# lsblk /dev/sdd1
NAME MAJ:MIN RM   SIZE RO TYPE MOUNTPOINTS
sdd1   8:49   1 103,3G  0 part /nix/store
                               /
```

Running `strace -f -o strace.log nix-channel --update` reveals that indeed, a child process of `nix-channel` is indeed encountering a "space left" error, returning from a system call. This time, the file and the operation seems to be different:

```
# strace -f -o strace.log nix-channel --update
unpacking channels...
error: opening lock file '/nix/var/nix/temproots/1464': No space left on device
error: program '/nix/store/4mkr15glvxawmvmadj7vhxpqndwb6rpg-nix-2.11.1/bin/nix-env' failed with exit code 1

# rg -C 2 ENOSPC strace.log
3858-1464  fcntl(8, F_SETLK, {l_type=F_UNLCK, l_whence=SEEK_SET, l_start=123, l_len=1}) = 0
3859-1464  newfstatat(AT_FDCWD, "/nix/var/nix/temproots/1464", 0x7ffc283070e0, AT_SYMLINK_NOFOLLOW) = -1 ENOENT (No such file or directory)
3860:1464  openat(AT_FDCWD, "/nix/var/nix/temproots/1464", O_RDWR|O_CREAT|O_CLOEXEC, 0600) = -1 ENOSPC (No space left on device)
3861-1464  newfstatat(AT_FDCWD, "/nix/var/nix/db/db.sqlite", {st_mode=S_IFREG|0644, st_size=43986944, ...}, 0) = 0
3862-1464  fcntl(6, F_SETLK, {l_type=F_WRLCK, l_whence=SEEK_SET, l_start=1073741824, l_len=1}) = 0
```

I'm at loss how to debug further. It was suggested in the matrix chat that the issue might be with temp file system, or with mounts in namespaces that are different from the global `/nix` or `/tmp`, but I don't know how to test these.

**Steps To Reproduce**

I can reproduce this on my own system by running `nix-channel --update`, but it seems hard to reproduce in general as the cause of the problem is unclear. Also, the exact reproduction doesn't seem to be deterministic, the file and the operation that fails is not always the same. Generally the error turns out to be "No space left on device", however.

**Expected behavior**

I expect the channel update to succeed, as there seems to be enough space. At least I expect to see a error message that clearly communicates the problem.

**`nix-env --version` output**

```
# nix-env --version
nix-env (Nix) 2.11.1
```

**Additional context**

Some other services seem to fail with similar errors too. For example:

```
Jul 24 11:44:27 mame acme-bitwarden.drasa.eu-start[1174]: /nix/store/p0zn13s844i1k18cmg8byrd9n9a14h9r-unit-script-acme-bitwarden.drasa.eu-start/bin/acme-bitwarden.drasa.eu-start: line 45: domainhash.txt: No space left on device
```

**Priorities**

Add :+1: to [issues you find important](https://github.com/NixOS/nix/issues?q=is%3Aissue+is%3Aopen+sort%3Areactions-%2B1-desc).

**Comment by golddranks** [comment]
More context: Even
```
# nix-env --list-generations
error: creating directory '/nix/var/nix/temproots': No space left on device
```
fails.

**Comment by golddranks** [comment]
Okay, problem solved. I'll try to detail the cause here to help out other people.

It was about inodes; inode exhaustion. `df -h` shows that there's still space, but `df -i` shows that 100% of inodes for `/` are consumed, so there are simply too much files.

I think that it's rather easy to arrive to this kind of situation with NixOS, because the different generations (and channel generations) contain simply a lot of tiny files. In my case, `system.autoUpgrade.enable = true;` was the culprit of accumulating generations. Turns out that `nix.gc.automatic = true;` isn't enough to remedy this.

There's a few personal takeaways:
- I didn't realize there were a huge amount of generations, because I didn't know that the system profile and the root profile are separate. I used to do `nix-env --list-generations` on root and wrongly conclude that the system was clean, but that doesn't show the system generations! One needs to do `nix-env --list-generations -p /nix/var/nix/profiles/system`.
- However, besides the system generations, there are also "channel generations", and these are being managed within the root profile. These aren't shown by default, either. To see these, do `nix-env --list-generations -p /nix/var/nix/profiles/per-user/root/channels`.
- These both accumulate, and consume a huge amount of inodes!
- Always when using `nix.gc.automatic = true;`, also turn on `nix.gc.options = "--delete-older-than 30d";`. Turns out this is very badly documented! Hopefully people that google this issue will find it at least here.

I think that there are also some takeaways for the NixOS project:
- Make `nix.gc.options = "--delete-older-than 30d";` default when `nix.gc.automatic = true;`!! Or at least make it a documented option in https://search.nixos.org/options?query=nix.gc
- NixOS store easily accumulates a lot of tiny files, so it seems to be more prone to inode exhaustion than your run-of-the mill OS! The tools could do well if they were able to catch this situation, and at least report it, because the standard "No space left on device" is a head-scratcher.
